domain/consensus/processes/pruningmanager/pruningmanager.go

```diff
@@ -468,6 +466,88 @@ func (pm *pruningManager) validateUTXOSetFitsCommitment(stagingArea *model.Stagi
 	return nil
 }
 
+func (pm *pruningManager) calculateDiffBetweenOldAndNewPruningPoints(stagingArea *model.StagingArea, newPruningHash *externalapi.DomainHash) (externalapi.UTXODiff, error) {
+	onEnd := logger.LogAndMeasureExecutionTime(log, "pruningManager.calculateDiffBetweenOldAndNewPruningPoints")
+	defer onEnd()
+	if newPruningHash.Equal(pm.genesisHash) {
+		return utxo.NewUTXODiff(), nil
+	}
+
+	oldPruningHash, err := pm.pruningStore.OldPruningPoint(pm.databaseContext, stagingArea)
+	if err != nil {
+		return nil, err
+	}
+	newPruningGhostDAG, err := pm.ghostdagDataStore.Get(pm.databaseContext, stagingArea, newPruningHash)
+	if err != nil {
+		return nil, err
+	}
+	oldPruningGhostDAG, err := pm.ghostdagDataStore.Get(pm.databaseContext, stagingArea, oldPruningHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// do the quick restore.
+	newPruningCurrentDiffChild := newPruningHash
+	oldPruningCurrentDiffChild := oldPruningHash
+	// We need to use BlueWork because it's the only thing that's monotonic in the whole DAG
+	newPruningCurrentDiffChildBlueWork := newPruningGhostDAG.BlueWork()
+	oldPruningCurrentDiffChildBlueWork := oldPruningGhostDAG.BlueWork()
+
+	var oldDiffs []externalapi.UTXODiff
+	var newDiffs []externalapi.UTXODiff
+	for {
+		// if newPruningCurrentDiffChildBlueWork > oldPruningCurrentDiffChildBlueWork
+		if newPruningCurrentDiffChildBlueWork.Cmp(oldPruningCurrentDiffChildBlueWork) == 1 {
+			utxoDiff, err := pm.utxoDiffStore.UTXODiff(pm.databaseContext, stagingArea, oldPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			oldDiffs = append(oldDiffs, utxoDiff)
+			oldPruningCurrentDiffChild, err = pm.utxoDiffStore.UTXODiffChild(pm.databaseContext, stagingArea, oldPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			diffChildGhostDag, err := pm.ghostdagDataStore.Get(pm.databaseContext, stagingArea, oldPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			oldPruningCurrentDiffChildBlueWork = diffChildGhostDag.BlueWork()
+		} else if newPruningCurrentDiffChild.Equal(oldPruningCurrentDiffChild) {
+			break
+		} else {
+			utxoDiff, err := pm.utxoDiffStore.UTXODiff(pm.databaseContext, stagingArea, newPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			newDiffs = append(newDiffs, utxoDiff)
+			newPruningCurrentDiffChild, err = pm.utxoDiffStore.UTXODiffChild(pm.databaseContext, stagingArea, newPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			diffChildGhostDag, err := pm.ghostdagDataStore.Get(pm.databaseContext, stagingArea, newPruningCurrentDiffChild)
+			if err != nil {
+				return nil, err
+			}
+			newPruningCurrentDiffChildBlueWork = diffChildGhostDag.BlueWork()
+		}
+	}
+	oldDiff := utxo.NewMutableUTXODiff()
+	for i := len(oldDiffs) - 1; i >= 0; i-- {
```

Comment:
Please explain in a comment why these diffs need to be applied in reverse order.